src/lib/models/emulation/Core.cc

```diff
@@ -7,35 +7,45 @@ namespace models {
 namespace emulation {
 
 // TODO: Expose as config option
-/** The number of bytes fetched each cycle. */
-const uint8_t FETCH_SIZE = 4;
 const unsigned int clockFrequency = 2.5 * 1e9;
 
 Core::Core(MemoryInterface& instructionMemory, MemoryInterface& dataMemory,
-           uint64_t entryPoint, uint64_t programByteLength,
            const arch::Architecture& isa)
     : instructionMemory_(instructionMemory),
       dataMemory_(dataMemory),
-      programByteLength_(programByteLength),
       isa_(isa),
-      pc_(entryPoint),
       registerFileSet_(isa.getRegisterFileStructures()),
-      architecturalRegisterFileSet_(registerFileSet_) {
-  // Pre-load the first instruction
-  instructionMemory_.requestRead({pc_, FETCH_SIZE});
-
-  // Query and apply initial state
-  auto state = isa.getInitialState();
-  applyStateChange(state);
-}
+      architecturalRegisterFileSet_(registerFileSet_) {}
 
 void Core::tick() {
   ticks_++;
+  isa_.updateSystemTimerRegisters(&registerFileSet_, ticks_);
 
-  if (hasHalted_) return;
+  switch (status_) {
+    case CoreStatus::idle:
+      idle_ticks_++;
+      return;
+    case CoreStatus::switching:
+      // Ensure no instructions left to execute, and not mid exception
```

Comment:
*Ensure there are no instructions left to execute and there's no active exception before context switching.